# pXg: proteomics X genomics
---
- [About pXg](#about-pxg)
- [Usage](#usage)
  - [Input format](#input-format)
  - [Output format](#output-format)
  - [Command-line interface](#command-line-interface)
- [3rd-party application](#3rd-party-application)
  - [IGV viewer](#igv-viewer)
  - [Percolator](#percolator)
---

## About pXg

pXg (proteomics X genomics), a software tool that enables the reliable identification of both cMAPs and ncMAPs from de novo peptide sequencing by utilizing RNA-Seq data.
<br>

## Usage
pXg can be integrated with any search engines such as PEAKS and pNovo3. 
It was developed for the reliable identification of ncMAPs from de novo peptide sequencing; however, it can also be used to capture the number of reads mapped to a peptide sequence.
### Input format
|Input | Description    | Format    | Requirement   |
| :---:   | :---:   | :---: | :---: |
| |It must be the same file used in the read alignment (e.g. STAR2)|GTF      | Yes   |
|Searh result  || TSV or CSV | Yes   |
| |GTF | It must be the same file used in the read alignment (e.g. STAR2)    | Yes   |
| |GTF | It must be the same file used in the read alignment (e.g. STAR2)    | Yes   |
### Output format
### Command-line interface
## 3rd-party application
### IGV viewer
### Percolator

